### ISSUE TITLE
Connection password displayed as clear text

### ISSUE THREAD

**Issue by xypron** [gotthardp/lorawan-server#538]
Passwords for connections should be entered via password field. They should not be transferred back to the GUI. The input should be of type = password.

https://docs.angularjs.org/api/ng/directive/input

**Comment by Tamburasca** [comment]
Agree with xypron: PW need to be hidden. Cheerio, Ralf A.

**Comment by altishchenko** [comment]
@Tamburasca Don't see any reason for this. As I commented in another discussion around this - it is not a _user_ interface - users should use some other App. Password visibility helps in troubleshooting a lot. Why not then hide the keys? Or DevEUIs? I can imagine the hell trying to get Erlang access to mnesia database to verify that the keys or passwords entered in the server are correct Once again, there should be _no_ user access to the server's web interface - that is what API is for.

**Comment by xypron** [comment]
@altishchenko @gotthardp 

Your arguments may be valid in a small company where there is only a single administrator. If you are working in a large organization things look different. Here the runtime support is often handed out to subcontractors. That you trust them to monitor your system does not imply that you are allowed to share passwords with all of their employees.

Essentially in larger organizations you may find regulations that would not allow to use your software if it displays the passwords in clear.

When you logon to Windows or Linux you can enter your password but it is not foreseen that you can display the password with which you logged on.

If you create a connection in a SAP system you once enter passwords for the connection but they will never be displayed to you.

This is simply the safety net that is needed when multiple persons with administration rights have access to a system.

**Comment by altishchenko** [comment]
@xypron I am sorry, but I do disagree with you, especially for big companies. I should make myself a little bit clearer. The problem is not in the server's web-admin, it is in usage - web-admin is not a product, it is a convenience, it is not at all end-user, device-admin interface. This interface _is_ Ok for the very small companies or home use ONLY.
My organization is not a small one (see www.softline.com), so we use another approach:
* There is a small number of admins, including myself, that manage the connectors and such and have full server access (we use scopes, do you?), nobody else has access to the web-admin at all (and to be honest we tend to automate things too out of laziness).
* Admins that manage devices (nodes, gateways and such) do so from _another_ application that uses REST API with the network server(s) to set things up, monitor, manage, etc...
* There is a **fixed** number of Profiles, that make an association between a device and handler/connector/etc. Anything that is needed for a devices user (our end user!) - attach the device to a proper Profile in a nice friendly GUI, with just one http call towards the server. And nothing apart from the useful stuff - "my devices", "my gateways", etc - is displayed at all.
* Our whole system is distributed, clusterized and structured around message brokers.
* If there is a new _type_ of devices coming up, our group deals with it first, setting up handlers, rabbitMQ queues for user access, Profiles, etc. This happens not too often now. Our changes are visible in the management application automatically.
* More often then not, we use web-admin for quick troubleshooting mostly. And with quick troubleshooting:
- hiding keys, passwords, connector details - is just one more hell to deal with.
- getting _http/https_ access over well known port is simple and easy.
- Getting -remsh access to a remote on-premises system is triple hell:
 Local admins tend to close all network ports apart from a few well known, so trying to make an erl -remsh access (as was onced proposed) becomes an issue for a few hours/days - in big companies local admins will not open any strange/unknown ports for remote access without an approvement by some big boss, and if you plus time difference of, say, 6 hours - that is where the third hell comes from. And, some government run companies will not allow a service VPN into their network no matter the need.

As for the last sentence :) in a large office your next table neighbor is your colleague who knows this passwords himself :)

This is only MY use case. This only MY IMHO. @gotthardp may have other thoughts on this.

But, for what it is worth, @xypron , we use Git, we run a local modified version of the software, we sync to the master with those features and fixes we need, we modify code to suite our needs - it is just that simple. If your corporate rules do not allow password display - well, don't display them :) Unfortunately use cases and corporate rules are not 'one size fits all'.
And consider: removing password access from the API completely (as you proposed once) will 100% break backup/restore procedure. It is always funny to forget to modify users.json (it doesn't show passwords in API) after backup and import it to a new server - users are created without any password field and you do need remsh access to fix it, or import one more user with "pass":"..." in it.

As I said, all of this is IMHO.

**Comment by Tamburasca** [comment]
@altishchenko 

I have never ever seen any software that that suffices industry standards in which passwords are legible. Sorry to disagree, Ralf A.

**Comment by xypron** [comment]
@altishchenko 
I do not think this is a good place for vile language be it in Russian or English.

"Fork this repository if you want to get something usable" is a very unexpected recommendation. If you don't use this software in production as is, the suggested patch should not be of any relevance to you. It would be nice if @gotthardp could give his view.

**Comment by gotthardp** [comment]
Hello. If possible, I'd like to avoid forks. There are not so many users nor contributors, so losing even a single one would be a pity.
The design intention was to design something small, that is easy to deploy and operate. The server was (and still is) designed for single-network deployments. Large or small, but single-network. Any device can be handled by any gateway. Any admin can modify (or even delete) any gateway or any device, so the "write-access" admins must be part of one team. Who needs multiple separated networks should deploy multiple separeted servers. The design intention however was not to design an insecure software. Voilating good security practices is wrong and should be avoided.

I can see there several features that can be implemented:
 * Hiding passwords (and other keys) because a non-admin may be behind my back. This must however be accompanied by an ability to display a plain-text password (the "eye" icon). As an admin I need to be able to read which password I entered.
 * Separated admin scope to manipulate passwords. If there are cubcontracted admins that cannot see the passwords, why don't we create a scope to view/edit the connectors and another scope to view/edit the passwords. Or something similar. That's what the scopes are for.

What is the residual security risk?
 * The passwords will be stored in a database. Well, they have to be anyway, because the passwords need to be provided when a connection is established.
 * The passwords will be transmitted to the web-admin app. The link is encrypted, so I wonder if it's really a vulnerability.

**Comment by altishchenko** [comment]
> @altishchenko
> I do not think this is a good place for vile language be it in Russian or English.

@xypron Sorry Heinrich, a bit of a stressfull day. It won't happen again.

The reason for my 'hot' reply was seeing a real danger in the issue. As I said, no user (in our case) has any access to the web-admin, only rightful admins do, so hiding passwords and keys for us will be one more pain to overcome and not so much of a security improvement.

While the server was designed to be compact and easy to use, it is quite successfull in big installations, like in our case, and we use it in cluster mode. All end-user configuration is done through another application with scopes used heavily.

> "Fork this repository if you want to get something usable" is a very unexpected recommendation. If you don't use this software in production as is, the suggested patch should not be of any relevance to you. It would be nice if @gotthardp could give his view.

@xypron Heinrich, some things can't be used as is. And when you are continuosly creating/updating/upgrading servers, small configuration bits may become an issue to maintain, so we fix them in our repo before the package is built at all, like db_master setting (we use clusters), network ports for packet forwarder and web-admin, certificate paths, frames_before_adr tuning, etc.

@gotthardp The **fork** I keep in my GitHub account is used only for pull requests when they are ready (such as, upcoming mysql support, when I finish testing it - unfortunately the orignal code is not maintained any more).

Our local Git repository contains all the tiny changes that we need to maintain our production installation and some experimental changes to the ADR algorithm for the devices with high mobility (they are not viable for pull request, as they are quite specific and we don't have the mobility indication in the server yet - thoughts for todo or may be not). Local 'devel' branch contains other sandboxed experiments, like NewChannelReq command implementation attempts, mysql, etc.

This type of 'forking' I kept in mind, when suggested keeping your own/local repository.

Sorry for the mess, always yours
Alex

P.S.: And again, I am sorry. I didn't intend to offend anyone!
